### ISSUE TITLE
s3: Please add Server-Side Encryption with Customer-Provided Encryption Keys (SSE-C)

### ISSUE THREAD

**Issue by arno01** [rclone/rclone#2827]
Please add S3 Server-Side Encryption with Customer-Provided Encryption Keys (SSE-C)

Minio provider should also be configurable to use the SSE-C.

#### What is your current rclone version (output from `rclone version`)?

```
rclone v1.45
- os/arch: android/arm64
- go version: go1.11.2
```

#### What problem are you are trying to solve?

Allow the customer to set their own encryption keys.
This is especially useful when Minio is used since it does not have KMS.

#### How do you think rclone should be changed to solve that?

https://docs.aws.amazon.com/AmazonS3/latest/dev/ServerSideEncryptionCustomerKeys.html

SSE-C supported in:
- Minio
  - https://github.com/minio/cookbook/blob/master/docs/how-to-use-minio-server-side-encryption-with-aws-cli.md
  - https://github.com/minio/mc/search?q=server+side+encryption&type=Code
- AWS-CLI SSE-C
  - https://github.com/aws/aws-cli/commit/4f798bf2024b40bb12ffa031cb651f4decfc6b49

**Comment by ncw** [comment]
It looks like this would be reasonably easy to add

Keys like this

```
SSECustomerAlgorithm: aws.String("AES256"), // header x-amz-server-side-encryption-customer-algorithm
   SSECustomerKey: aws.String("czn8qrbUsT/5y5Hr2i93ImWmIQLCZLOL"), // header x-amz-server-side-encryption-customer-key
   SSECustomerKeyMD5: aws.String("ME4ss65LcXQBY2CynVdZyA=="), //header x-amz-server-side-encryption-customer-key-MD5
```

would need to go here

https://github.com/ncw/rclone/blob/e7684b7ed5d2b325c2ce7790e8c0a663cc0a870b/backend/s3/s3.go#L1643

And

https://github.com/ncw/rclone/blob/e7684b7ed5d2b325c2ce7790e8c0a663cc0a870b/backend/s3/s3.go#L1669

I'm not sure if server side copy would need that too - probably.

Want to have a go?

**Comment by Boxfort** [comment]
I need this for Ceph so I'm taking a shot at implementing!

**Comment by ncw** [comment]
I've fixed up the remaining things for the SSE-C implementation

Can you have a go with this?

[v1.54.0-beta.4905.cbd93519c.fix-s3-sse](https://beta.rclone.org/branch/fix-s3-sse/v1.54.0-beta.4905.cbd93519c.fix-s3-sse/) on branch [fix-s3-sse](https://github.com/rclone/rclone/tree/fix-s3-sse) (uploaded in 15-30 mins)

Note that you don't need to set `--s3-sse-customer-key-md5` any more - rclone will calculate that for you.

**Comment by ncw** [comment]
Update which stored MD5 as metadata on SSE uploaded objects (like multipart uploads have)

[v1.54.0-beta.4907.78e3ba830.fix-s3-sse](https://beta.rclone.org/branch/fix-s3-sse/v1.54.0-beta.4907.78e3ba830.fix-s3-sse/) on branch [fix-s3-sse](https://github.com/rclone/rclone/tree/fix-s3-sse) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54
